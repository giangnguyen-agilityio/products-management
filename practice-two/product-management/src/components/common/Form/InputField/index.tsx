import React, { memo } from 'react'
import {
  FormControl,
  FormLabel,
  FormErrorMessage,
  Input,
} from '@chakra-ui/react'

interface InputFieldProps {
  variant?: 'input' | 'textarea'
  name: string
  label: string
  value: string | number
  onChange: (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void
  type?: string
  errorMessage?: string
}

const InputField: React.FC<InputFieldProps> = ({
  variant = 'input',
  name,
  label,
  value,
  onChange,
  type = 'text',
  errorMessage,
}) => (
  <FormControl className="input-field" isRequired>
    <FormLabel
      className="form-label"
      fontFamily="Oswald-Regular"
      fontSize="xl"
      letterSpacing="1px"
    >
      {label}:
    </FormLabel>
    <Input
      as={variant}
      border="2px solid"
      _hover={{ borderColor: 'primary' }}
      _focus={{ borderColor: 'primary' }}
      borderColor="gray.300"
      backgroundColor="background"
      padding={2}
      borderRadius="6px"
      height={variant === 'textarea' ? '100px' : '45px'}
      type={type}
      name={name}
      value={value}
      onChange={onChange}
    />
    <FormErrorMessage>{errorMessage}</FormErrorMessage>
  </FormControl>
)

export default memo(InputField)
