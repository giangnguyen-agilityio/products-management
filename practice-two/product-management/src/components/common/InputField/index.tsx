// Libraries
import React, { InputHTMLAttributes, memo } from 'react'
import {
  FormControl,
  FormErrorMessage,
  FormLabel,
  Input,
} from '@chakra-ui/react'

// Define the props interface for the InputField component
interface InputFieldProps extends InputHTMLAttributes<HTMLInputElement> {
  variant?: 'input' | 'textarea'
  name: string
  label: string
  value?: string | number
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void
  type?: string
  errorMessage?: string
  isReadOnly?: boolean
  min?: number
  max?: number
}

// Custom deep comparison function for props
export function arePropsEqual(
  prevProps: InputFieldProps,
  nextProps: InputFieldProps
) {
  return (
    prevProps.variant === nextProps.variant &&
    prevProps.name === nextProps.name &&
    prevProps.label === nextProps.label &&
    prevProps.value === nextProps.value &&
    prevProps.errorMessage === nextProps.errorMessage &&
    prevProps.isReadOnly === nextProps.isReadOnly &&
    prevProps.min === nextProps.min &&
    prevProps.max === nextProps.max
  )
}

const InputField: React.FC<InputFieldProps> = ({
  variant = 'input',
  name,
  label,
  value,
  onChange,
  type = 'text',
  errorMessage,
  isReadOnly,
  min,
  max,
}) => (
  <FormControl className="input-field" isRequired isInvalid={!!errorMessage}>
    {/* Form Label */}
    <FormLabel
      className="form-label"
      fontFamily="Oswald-Regular"
      fontSize="mediumLarge"
      letterSpacing="1px"
    >
      {label}:
    </FormLabel>

    {/* Input field */}
    <Input
      as={variant}
      readOnly={isReadOnly}
      variant="primary"
      padding={2}
      height={variant === 'textarea' ? '100px' : '45px'}
      type={type}
      name={name}
      value={value}
      onChange={onChange}
      min={min}
      max={max}
      aria-label="Input Field"
    />

    {/* Error message */}
    {errorMessage && (
      <FormErrorMessage color="error">{errorMessage}</FormErrorMessage>
    )}
  </FormControl>
)

export default memo(InputField, arePropsEqual)
