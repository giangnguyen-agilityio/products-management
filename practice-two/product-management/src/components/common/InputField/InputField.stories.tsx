import type { Meta, StoryObj } from '@storybook/react'
import InputField from './index'
const meta = {
  title: 'Practice Two/InputField Component',
  component: InputField,
  decorators: [
    Story => (
      <div
        style={{
          padding: '10px',
        }}
      >
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} satisfies Meta<typeof InputField>
export default meta
type Story = StoryObj<typeof meta>
export const TextInput: Story = {
  args: {
    label: 'Name of product',
    name: 'productName',
    type: 'text',
    errorMessage: 'Please enter the name of the product',
  },
}
export const UploadImageInput: Story = {
  args: {
    label: 'Upload Your Image',
    name: 'productImage',
    type: 'file',
    accept: 'image/*',
    errorMessage: 'Please select the image to upload',
  },
}
export const NumberInput: Story = {
  args: {
    label: 'Enter your price',
    name: 'productPrice',
    type: 'number',
    min: 0,
    errorMessage: 'The value must be greater than 0',
  },
}
