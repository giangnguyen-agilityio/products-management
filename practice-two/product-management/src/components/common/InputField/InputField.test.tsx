import { render, screen, fireEvent } from '@testing-library/react'
import InputField, { arePropsEqual } from './'

describe('InputField Component', () => {
  const defaultProps = {
    name: 'test',
    label: 'Test Label',
    value: 'initial value',
    onChange: jest.fn(),
    errorMessage: 'Error message',
    isReadOnly: false,
    min: 0,
    max: 10,
  }

  it('renders the component with default props', () => {
    const { container, getByLabelText } = render(
      <InputField {...defaultProps} />
    )
    const input = getByLabelText('Input Field')
    expect(input).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('renders an error message when errorMessage prop is provided', () => {
    const { getByText } = render(
      <InputField {...defaultProps} errorMessage="This is an error" />
    )
    const errorMessage = getByText('This is an error')
    expect(errorMessage).toBeInTheDocument()
  })

  it('calls the onChange function when input value changes', () => {
    const { getByLabelText } = render(<InputField {...defaultProps} />)
    const input = getByLabelText('Input Field')

    fireEvent.change(input, { target: { value: 'New Value' } })
    expect(defaultProps.onChange).toHaveBeenCalledTimes(1)
  })

  it('renders a textarea when variant is set to "textarea"', () => {
    const { getByLabelText } = render(
      <InputField {...defaultProps} variant="textarea" />
    )
    const textarea = getByLabelText('Input Field')
    expect(textarea.tagName.toLowerCase()).toBe('textarea')
    expect(textarea).toBeInTheDocument()
    expect(textarea).toHaveStyle('height: 100px')
  })

  it('renders error message when errorMessage prop is provided', () => {
    render(<InputField {...defaultProps} />)
    const errorMessage = screen.getByText(defaultProps.errorMessage)
    expect(errorMessage).toBeInTheDocument()
  })

  it('renders with correct input type', () => {
    render(<InputField {...defaultProps} type="number" />)
    const input = screen.getByLabelText('Input Field')
    expect(input).toHaveAttribute('type', 'number')
  })

  it('displays readOnly input when isReadOnly prop is true', () => {
    render(<InputField {...defaultProps} isReadOnly />)
    const input = screen.getByLabelText('Input Field')
    expect(input).toHaveAttribute('readonly')
  })

  it('renders input with min and max attributes when provided', () => {
    render(<InputField {...defaultProps} />)
    const input = screen.getByLabelText('Input Field')
    expect(input).toHaveAttribute('min', defaultProps.min.toString())
    expect(input).toHaveAttribute('max', defaultProps.max.toString())
  })

  it('does not render error message when errorMessage prop is not provided', () => {
    render(<InputField {...defaultProps} errorMessage="" />)
    const errorMessage = screen.queryByText(defaultProps.errorMessage)
    expect(errorMessage).toBeNull()
  })

  it('should return true for equal props', () => {
    const prevProps = {
      name: 'username',
      label: 'Username',
      value: 'testuser',
      errorMessage: 'Invalid input',
      isReadOnly: false,
      min: 0,
      max: 10,
      onChange: () => {},
    }

    const nextProps = { ...prevProps }

    expect(arePropsEqual(prevProps, nextProps)).toBe(true)
  })

  it('should return false for different props', () => {
    const prevProps = {
      name: 'username',
      label: 'Username',
      value: 'testuser',
      errorMessage: 'Invalid input',
      isReadOnly: false,
      min: 0,
      max: 10,
      onChange: () => {},
    }

    const nextProps = {
      ...prevProps,
      value: 'updateduser',
    }

    expect(arePropsEqual(prevProps, nextProps)).toBe(false)
  })
})
