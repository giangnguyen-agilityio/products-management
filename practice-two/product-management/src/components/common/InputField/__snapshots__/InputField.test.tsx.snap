// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField Component renders the component with default props 1`] = `
<div>
  <div
    class="chakra-form-control input-field css-0"
    role="group"
  >
    <label
      class="chakra-form__label form-label css-k0s0px"
      data-invalid=""
      for="field-:r0:"
      id="field-:r0:-label"
    >
      Test Label
      :
      <span
        aria-hidden="true"
        class="chakra-form__required-indicator css-0"
        role="presentation"
      >
        *
      </span>
    </label>
    <input
      aria-describedby="field-:r0:-feedback"
      aria-invalid="true"
      aria-label="Input Field"
      aria-required="true"
      class="chakra-input css-16crra2"
      id="field-:r0:"
      max="10"
      min="0"
      name="test"
      required=""
      type="text"
      value="initial value"
    />
    <div
      aria-live="polite"
      class="chakra-form__error-message css-c80spl"
      id="field-:r0:-feedback"
    >
      Error message
    </div>
  </div>
</div>
`;
